CRM/Custom/Form/CustomDataTrait.php

```diff
@@ -1 +1,92 @@
 <?php
+/*
+ +--------------------------------------------------------------------+
+ | Copyright CiviCRM LLC. All rights reserved.                        |
+ |                                                                    |
+ | This work is published under the GNU AGPLv3 license with some      |
+ | permitted exceptions and without any warranty. For full license    |
+ | and copyright information, see https://civicrm.org/licensing       |
+ +--------------------------------------------------------------------+
+ */
+
+/**
+ *
+ * @package CRM
+ * @copyright CiviCRM LLC https://civicrm.org/licensing
+ */
+
+/**
+ * This trait supports adding custom data to forms.
+ *
+ * @internal
+ */
+trait CRM_Custom_Form_CustomDataTrait {
+
+  /**
+   * Add custom data fields to the form.
+   *
+   * This function takes responsibility for registering the appropriate fields with QuickForm.
+   *
+   * It does not assign variables to the smarty layer to assist with rendering or setDefaults.
+   * Those additional steps are not required on the 'main form' when the CustomDataForm is
+   * being rendered by ajax as that process takes care of the presentation.
+   *
+   * However, the fields need to be registered with QuickForm to ensure that on submit
+   * they are 'accepted' by quick form. Quick form validates the contents of $_POST
+   * and only fields that have been registered make it through to the values
+   * available in `$form->getSubmittedValues()`.
+   *
+   * @param string $entity
+   * @param array $filters
+   *
+   * @throws \CRM_Core_Exception
+   */
+  protected function addCustomDataFieldsToForm(string $entity, array $filters = []): void {
+    $fields = civicrm_api4($entity, 'getFields', [
+      'action' => 'create',
+      'values' => $filters,
+      'where' => [
+        ['type', '=', 'Custom'],
```

Comment:
I think forms are supposed to exclude the `is_view` fields which would be done like this:
```suggestion
        ['type', '=', 'Custom'],
        ['readonly', '=', FALSE],
```